Complaints about TERM environment variable not set in tests

If a user has for instance `clear` in e.g. `.bash_profile`, this appears running `verdi devel tests`:

`AssertionError: 'TERM environment variable not set.\n' != ''`

When running the `ssh` tests.

How strict should Aiida be with respect to the system config files?  Commands like clear in your `.bash_profile` etc. should possibly be allowed (albeit not always a smart thing to do). Possibly, the `-t` option can be used. Or checking for this error and ignoring it. If neither it should maybe be mentioned explicitly in the documentation? There is of course other system based workarounds to this which is not related to Aiida, like the obvious, removing or moving the clear statement.

--- a/aiida/cmdline/commands/cmd_computer.py
+++ b/aiida/cmdline/commands/cmd_computer.py
@@ -68,6 +68,58 @@ def _computer_test_get_jobs(transport, scheduler, authinfo):  # pylint: disable=
     return True
 
 
+def _computer_test_no_unexpected_output(transport, scheduler, authinfo):  # pylint: disable=unused-argument
+    """
+    Test that there is no unexpected output from the connection.
+
+    This can happen if e.g. there is some spurious command in the
+    .bashrc or .bash_profile that is not guarded in case of non-interactive
+    shells.
+
+    :param transport: an open transport
+    :param scheduler: the corresponding scheduler class
+    :param authinfo: the AuthInfo object (from which one can get computer and aiidauser)
+    :return: True if the test succeeds, False if it fails.
+    """
+    # Execute a command that should not return any error
+    echo.echo("> Checking that no spurious output is present...")
+    retval, stdout, stderr = transport.exec_command_wait('echo -n')
+    if retval != 0:
+        echo.echo_error("* ERROR! The command 'echo -n' returned a " "non-zero return code ({})!".format(retval))
+        return False
+    if stdout:
+        echo.echo_error(u"""* ERROR! There is some spurious output in the standard output,
+that we report below between the === signs:
+=========================================================
+{}
+=========================================================
+Please check that you don't have code producing output in
+your ~/.bash_profile (or ~/.bashrc). If you don't want to
+remove the code, but just to disable it for non-interactive
+shells, see comments in issue #1980 on GitHub:
+https://github.com/aiidateam/aiida_core/issues/1890
+(and in the AiiDA documentation, linked from that issue)
+""".format(stdout))
+        return False
+    if stderr:
+        echo.echo_error(u"""* ERROR! There is some spurious output in the stderr,
+that we report below between the === signs:
+=========================================================
+{}
+=========================================================
+Please check that you don't have code producing output in
+your ~/.bash_profile (or ~/.bashrc). If you don't want to
+remove the code, but just to disable it for non-interactive
+shells, see comments in issue #1980 on GitHub:
+https://github.com/aiidateam/aiida_core/issues/1890
+(and in the AiiDA documentation, linked from that issue)
+""".format(stderr))
+        return False
+
+    echo.echo("      [OK]")
+    return True
+
+
 def _computer_create_temp_file(transport, scheduler, authinfo):  # pylint: disable=unused-argument
     """
     Internal test to check if it is possible to create a temporary file
@@ -506,7 +558,7 @@ def computer_test(user, print_traceback, computer):
         with trans:
             sched.set_transport(trans)
             num_tests += 1
-            for test in [_computer_test_get_jobs, _computer_create_temp_file]:
+            for test in [_computer_test_no_unexpected_output, _computer_test_get_jobs, _computer_create_temp_file]:
                 num_tests += 1
                 try:
                     succeeded = test(transport=trans, scheduler=sched, authinfo=authinfo)

--- a/docs/source/install/details/troubleshooting.rst
+++ b/docs/source/install/details/troubleshooting.rst
@@ -3,7 +3,10 @@
 Troubleshooting
 ===============
 
-* On a clean Ubuntu 16.04 install the pip install command ``pip install -e aiida_core``
+Installation phase
+------------------
+
+* [**numpy dependency**] On a clean Ubuntu 16.04 install the pip install command ``pip install -e aiida_core``
   may fail due to a problem with dependencies on the ``numpy`` package. In this case
   you may be presented with a message like the following::
 
@@ -20,57 +23,7 @@ Troubleshooting
 
   This should fix the dependency error.
 
-* If the ``pip install`` command gives you an error that resembles the one
-  shown below, you might need to downgrade to an older version of pip::
-
-    Cannot fetch index base URL https://pypi.python.org/simple/
-
-  To downgrade pip, use the following command::
-
-    sudo easy_install pip==1.2.1
-
-* In order to use the AiiDA objects and functions in Jupyter, this latter has to be instructed to use the iPython kernel installed in the AiiDA virtual environment. This happens by default if you install AiiDA with ``pip`` including the ``notebook`` option and run Jupyter from the AiiDA virtual environment.
-
-  If, for any reason, you do not want to install Jupyter in the virtual environment, you might consider to install it out of the virtual environment, if not already done::
-
-      $ pip install jupyter
-
-  Then, activate the AiiDA virtual environment::
-
-      $ source ~/<aiida.virtualenv>/bin/activate
-
-  and setup the AiiDA iPython kernel::
-
-      $ pip install ipykernel
-      $ python -m ipykernel install --user --name=<aiida.kernel.name>
-
-  where you have chosen a meaningful name for the new kernel.
-
-  Finally, start a Jupyter server::
-
-      $ jupyter notebook
-
-  and from the newly opened browser tab select ``New -> <aiida.kernel.name>``
-
-
-* When installing the ``ssh_kerberos`` optional requirement through Anaconda you may encounter the following error on Ubuntu machines::
-
-    version 'GFORTRAN_1.4' not found (required by /usr/lib/libblas.so.3)
-
-  This is related to an open issue in anaconda `ContinuumIO/anaconda-issues#686`_.
-  A potential solution is to run the following command::
-
-    export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libgfortran.so.3
-
-.. _ContinuumIO/anaconda-issues#686: https://github.com/ContinuumIO/anaconda-issues/issues/686
-
-* Several users reported the need to install also ``libpq-dev`` (header files for libpq5 - PostgreSQL library)::
-
-    apt-get install libpq-dev
-
-  But under Ubuntu 12.04 this is not needed.
-
-* If the installation fails while installing the packages related
+* [**Database installation and location**] If the installation fails while installing the packages related
   to the database, you may have not installed or set up the database
   libraries.
 
@@ -103,8 +56,8 @@ Troubleshooting
 
 .. _Stackoverflow link: http://stackoverflow.com/questions/21079820/how-to-find-pg-config-pathlink
 
-
-* For some reasons, on some machines (notably often on Mac OS X) there is no
+* [**ensuring a UTF-8 locale**]For some reasons, on some machines 
+  (notably often on Mac OS X) there is no
   default locale defined, and when you run ``verdi setup`` for the first
   time it fails (see also `this issue`_ of django).
   Run in your terminal (or maybe even better, add to your ``.bashrc``, but
@@ -117,9 +70,77 @@ Troubleshooting
 
 .. _this issue: https://code.djangoproject.com/ticket/16017
 
+* [**possible Ubuntu dependencies**] Several users reported the need to install 
+  also ``libpq-dev`` (header files for libpq5 - PostgreSQL library)::
 
-* Within a virtual environment, attempt to visualize a structure with ``ase`` (either from the shell, or using the 
-  command ``verdi data structure show --format=ase <PK>``), might end up with the following error message::
+    apt-get install libpq-dev
+
+  But under Ubuntu 12.04 this is not needed.
+
+Configuring remote SSH computers
+--------------------------------
+
+* [**ssh_kerberos installation**] When installing the ``ssh_kerberos`` *optional*
+  requirement through Anaconda you may encounter the following error on Ubuntu machines::
+
+    version 'GFORTRAN_1.4' not found (required by /usr/lib/libblas.so.3)
+
+  This is related to an open issue in anaconda `ContinuumIO/anaconda-issues#686`_.
+  A potential solution is to run the following command::
+
+    export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libgfortran.so.3
+
+.. _ContinuumIO/anaconda-issues#686: https://github.com/ContinuumIO/anaconda-issues/issues/686
+
+* [**.bashrc and .bash_profile behavior on remote computers**] When connecting
+  to remote computers, AiiDA (like other codes as ``sftp``) might get very confused
+  if you have code in your ``.bashrc`` or ``.bash_profile`` producing output
+  or e.g. running commands like ``clean`` that require a terminal.
+
+  For instance, if you add a ``echo "a"`` in your .bashrc and then try to SFTP
+  a file from it, you will get an error like ``Received message too long 1091174400``.
+
+  If you still want to have code that needs an interactive shell (``echo``, 
+  ``clean``, ...), but you want to disable it for non-interactive shells, put your
+  code in a guard like this::
+
+    case $- in
+    *i*)
+       ## Put here code that can only work on interactive shells
+       clear
+       echo "Hi! This is a new shell"
+       ;;
+    esac
+
+  Also, if you need to have some environment variables set (e.g. the path
+  to an executable), make sure that this is set in the ``.bashrc`` file 
+  (and not only in the ``.bash_profile`` or equivalent files that are only
+  loaded for login shells; see e.g. this `StackExchange thread`_ for the difference 
+  - as suggested in one of the answers, you might want to put the PATH definitions
+  in ``.bashrc``, and source it from ``.bash_profile``).
+
+  To test if a command works in a non-interactive shell, you can run::
+
+     ssh <MACHINENAME> <COMMAND>
+
+  (this is equivalent to what AiiDA does) and check, e.g., that the executable
+  is found, that there is no spurious output, etc.
+
+  **Notes**: The fact that there is no spurious output is checked by
+  ``verdi computer test``, which suggests how to solve the problem.
+  You can track the discussion on this issue in `aiidateam/aiida_core#1890`_.
+
+.. _aiidateam/aiida_core#1890: https://github.com/aiidateam/aiida_core/issues/1890
+.. _StackExchange thread: https://apple.stackexchange.com/questions/51036/what-is-the-difference-between-bash-profile-and-bashrc
+
+
+Improvements for dependencies
+-----------------------------
+* [**Activating the ASE visualizer**] Within a virtual environment, 
+  attempt to visualize a structure 
+  with ``ase`` (either from the shell, or using the 
+  command ``verdi data structure show --format=ase <PK>``), 
+  might end up with the following error message::
   
      ImportError: No module named pygtk
 
@@ -153,7 +174,38 @@ Troubleshooting
 
   After that, ``verdi data structure show --format=ase <PK>`` should work.
 
-* [*Only for developers*] The developer tests of the *SSH* transport plugin are
+Use in ipython/jupyter
+----------------------
+
+* In order to use the AiiDA objects and functions in Jupyter, this latter has to be instructed to use the iPython kernel installed in the AiiDA virtual environment. This happens by default if you install AiiDA with ``pip`` including the ``notebook`` option and run Jupyter from the AiiDA virtual environment.
+
+  If, for any reason, you do not want to install Jupyter in the virtual environment, you might consider to install it out of the virtual environment, if not already done::
+
+      $ pip install jupyter
+
+  Then, activate the AiiDA virtual environment::
+
+      $ source ~/<aiida.virtualenv>/bin/activate
+
+  and setup the AiiDA iPython kernel::
+
+      $ pip install ipykernel
+      $ python -m ipykernel install --user --name=<aiida.kernel.name>
+
+  where you have chosen a meaningful name for the new kernel.
+
+  Finally, start a Jupyter server::
+
+      $ jupyter notebook
+
+  and from the newly opened browser tab select ``New -> <aiida.kernel.name>``
+
+.. _updating_aiida:
+
+For developers (testing)
+------------------------
+
+* [**Making the SSH tests pass**] The developer tests of the *SSH* transport plugin are
   performed connecting to ``localhost``. The tests will fail if
   a passwordless ssh connection is not set up. Therefore, if you want to run
   the tests:
@@ -171,7 +223,4 @@ Troubleshooting
 
     (it will ask your password, because it is connecting via ssh to ``localhost``
     to install your public key inside ~/.ssh/authorized_keys).
-
-.. _updating_aiida:
-
 

--- a/docs/source/install/details/troubleshooting.rst
+++ b/docs/source/install/details/troubleshooting.rst
@@ -92,42 +92,36 @@ Configuring remote SSH computers
 
 .. _ContinuumIO/anaconda-issues#686: https://github.com/ContinuumIO/anaconda-issues/issues/686
 
-* [**.bashrc and .bash_profile behavior on remote computers**] When connecting
+* [**.bashrc and .bash_profile behavior on remote computers**] 
+  (**NOTE** This applies also to a computer configured via a ``local`` transport!)
+  
+  When connecting
   to remote computers, AiiDA (like other codes as ``sftp``) might get very confused
   if you have code in your ``.bashrc`` or ``.bash_profile`` producing output
   or e.g. running commands like ``clean`` that require a terminal.
 
-  For instance, if you add a ``echo "a"`` in your .bashrc and then try to SFTP
+  For instance, if you add a ``echo "a"`` in your ``.bashrc`` and then try to SFTP
   a file from it, you will get an error like ``Received message too long 1091174400``.
 
   If you still want to have code that needs an interactive shell (``echo``, 
-  ``clean``, ...), but you want to disable it for non-interactive shells, put your
-  code in a guard like this::
+  ``clean``, ...), but you want to disable it for non-interactive shells, put 
+  at the top of your file a guard like this::
 
-    case $- in
-    *i*)
-       ## Put here code that can only work on interactive shells
-       clear
-       echo "Hi! This is a new shell"
-       ;;
-    esac
+    if [[ $- != *i* ]] ; then
+      # Shell is non-interactive.  Be done now!
+      return
+    fi
 
-  Also, if you need to have some environment variables set (e.g. the path
-  to an executable), make sure that this is set in the ``.bashrc`` file 
-  (and not only in the ``.bash_profile`` or equivalent files that are only
-  loaded for login shells; see e.g. this `StackExchange thread`_ for the difference 
-  - as suggested in one of the answers, you might want to put the PATH definitions
-  in ``.bashrc``, and source it from ``.bash_profile``).
+  Everything below this will not be executed in a non-interactive shell.
+  **Note**: Still, you might want to have some code on top, like e.g. setting the PATH or 
+  similar, if this needs to be run also in the case of non-interactive shells.
 
-  To test if a command works in a non-interactive shell, you can run::
+  To test if a the computer does not produce spurious output, run (after 
+  configuring)::
 
-     ssh <MACHINENAME> <COMMAND>
+     verdi computer test <COMPUTERNAME>
 
-  (this is equivalent to what AiiDA does) and check, e.g., that the executable
-  is found, that there is no spurious output, etc.
-
-  **Notes**: The fact that there is no spurious output is checked by
-  ``verdi computer test``, which suggests how to solve the problem.
+  which checks and, in case of problems, suggests how to solve the problem.
   You can track the discussion on this issue in `aiidateam/aiida_core#1890`_.
 
 .. _aiidateam/aiida_core#1890: https://github.com/aiidateam/aiida_core/issues/1890

